doghouse/appengine/github.go

```diff
@@ -0,0 +1,236 @@
+	const u = "https://github.com/login/oauth/access_token"
+	cli := urlfetch.Client(ctx)
+	data := url.Values{}
+	data.Set("client_id", g.clientID)
+	data.Set("client_secret", g.clientSecret)
+	data.Set("code", code)
+	data.Set("state", state)
+
+	req, err := http.NewRequest("POST", u, strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %v", err)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Accept", "application/vnd.github.machine-man-preview+json")
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := cli.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to request access token: %v", err)
+	}
+	defer res.Body.Close()
+
+	b, _ := ioutil.ReadAll(res.Body)
+
+	var token struct {
+		AccessToken string `json:"access_token"`
+	}
+	if err := json.NewDecoder(bytes.NewReader(b)).Decode(&token); err != nil {
+		return "", fmt.Errorf("failed to decode response: %v", err)
+	}
+
+	if token.AccessToken == "" {
+		log.Errorf(ctx, "response doesn't contain token (resopnse: %s)", b)
+		return "", errors.New("response doesn't contain GitHub access token")
+	}
+
+	return token.AccessToken, nil
+}
+
+func (g *GitHubHandler) token(r *http.Request) (bool, string) {
+	b, err := g.tokenStore.Get(r)
+	if err != nil {
+		return false, ""
+	}
+	return true, string(b)
+}
+
+func (g *GitHubHandler) HandleGitHubTop(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+
+	ok, token := g.token(r)
+	if !ok {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	ghcli := github.NewClient(NewAuthClient(ctx, urlfetch.Client(ctx).Transport, ts))
+	u, _, err := ghcli.Users.Get(ctx, "")
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "cannot get GitHub authenticated user")
+		return
+	}
+
+	fmt.Fprintf(w, "Hi, %s!\n", u.GetName())
+
+	{
+		fmt.Fprintln(w, "")
+		ghcli, err := server.NewGitHubClient(ctx, &server.NewGitHubClientOption{
+			Client:        urlfetch.Client(ctx),
+			IntegrationID: g.integrationID,
+			PrivateKey:    g.privateKey,
+		})
+		if err != nil {
+			fmt.Fprintln(w, err)
+			return
+		}
+		app, _, err := ghcli.Apps.Get(ctx, "")
+		if err != nil {
+			fmt.Fprintln(w, err)
+		} else {
+			fmt.Fprintf(w, "%s app setting: %s\n", app.GetName(), app.GetHTMLURL())
+		}
+	}
+
+	{
+		fmt.Fprintln(w, "")
+		fmt.Fprintf(w, "Installation setting\n")
+		installations, _, err := ghcli.Apps.ListUserInstallations(ctx, nil)
+		if err != nil {
+			fmt.Fprintln(w, err)
+			return
+		}
+		for _, inst := range installations {
+			fmt.Fprintf(w, "\t%s: %s\n", inst.GetAccount().GetLogin(), inst.GetHTMLURL())
+		}
+	}
+}
+
+func NewAuthClient(ctx context.Context, base http.RoundTripper, token oauth2.TokenSource) *http.Client {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function NewAuthClient should have comment or be unexported